src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,259 @@
+    if (newMap.DeviceName) {
+      const res: InstanceBlockDeviceMapping = {
+        instanceId: instance?.id ?? undefined,
+        volumeId: volume?.id ?? undefined,
+        instance,
+        volume,
+        deviceName: newMap.DeviceName,
+        region,
+        deleteOnTermination: newMap.Ebs?.DeleteOnTermination ?? true,
+      };
+      return res;
+    } else return undefined;
+  }
+
+  attachVolumeInternal = crudBuilder<EC2, 'attachVolume'>('attachVolume', (VolumeId, InstanceId, Device) => ({
+    VolumeId,
+    InstanceId,
+    Device,
+  }));
+
+  modifyBlockDeviceMapping = crudBuilder<EC2, 'modifyInstanceAttribute'>(
+    'modifyInstanceAttribute',
+    (InstanceId, BlockDeviceMappings) => ({
+      InstanceId,
+      BlockDeviceMappings,
+    }),
+  );
+
+  waitUntilAvailable(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.AVAILABLE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  detachVolumeInternal = crudBuilder<EC2, 'detachVolume'>('detachVolume', VolumeId => ({ VolumeId }));
+
+  detachVolume = async (client: EC2, VolumeId: string) => {
+    await this.detachVolumeInternal(client, VolumeId);
+    await this.waitUntilAvailable(client, VolumeId);
+  };
+
+  waitUntilInUse(client: EC2, volumeId: string) {
+    return this.module.generalPurposeVolume.volumeWaiter(client, volumeId, (vol: AWSVolume | undefined) => {
+      // If state is not 'in-use' retry
+      if (!Object.is(vol?.State, VolumeState.IN_USE)) {
+        return { state: WaiterState.RETRY };
+      }
+      return { state: WaiterState.SUCCESS };
+    });
+  }
+
+  attachVolume = async (client: EC2, VolumeId: string, InstanceId: string, Device: string) => {
+    const result = await this.attachVolumeInternal(client, VolumeId, InstanceId, Device);
+    if (result) await this.waitUntilInUse(client, VolumeId);
+    else return false;
+    return true;
+  };
+
+  cloud: Crud<InstanceBlockDeviceMapping> = new Crud({
+    create: async (es: InstanceBlockDeviceMapping[], ctx: Context) => {
+      const out: InstanceBlockDeviceMapping[] = [];
+      for (const e of es) {
+        // if we have different regions we may be in a multi-region update situation, just skip
+        if (e.region !== e.instance.region) continue;
+
+        // if instance is not created we are in the first step, no need to create anything
+        if (!e.instance.instanceId) continue;
+        const client = (await ctx.getAwsClient(e.region)) as AWS;
+
+        // read volume details
+        if (!e.volumeId) throw new Error('Cannot attach empty volumes to an instance already created');
+        if (e.volume) {
+          if (e.volume.region !== e.instance.region)
+            throw new Error('Cannot create a mapping between different regions');
```

Comment:
Is possible for this to happen? Isn't this handled by the database with the FKs?

Reply:
indeed it was an older code leftover